### PR TITLE
[cxx-interop] Support nested C++ record types.

### PR DESCRIPTION
**Pull request by zoecarver**
Simply returns "None" for C++ records in "getSpecialManglingContext" (same logic as namespaces) to prevent an assertion.

### CODE REVIEW

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: lib/AST/ASTMangler.cpp [hunk 1]
```diff
@@ -1813,7 +1813,13 @@ ASTMangler::getSpecialManglingContext(const ValueDecl *decl,
...
+        // In C, "nested" structs, unions, enums, etc. will become sibilings:
+        //   struct Foo { struct Bar { }; }; -> struct Foo { }; struct Bar { };
+        // Whereas in C++, nested records will actually be nested. So if this is
+        // a C++ record, simply treat it like a namespace and exit early.
+        if (isa<clang::NamespaceDecl>(clangDC) ||
+            isa<clang::CXXRecordDecl>(clangDC))
+          return None;
         assert(clangDC->getRedeclContext()->isTranslationUnit() &&
                "non-top-level Clang types not supported yet");
         (void)clangDC;
```
Comment (varungandhi-apple) [inline lib/AST/ASTMangler.cpp:1825]: Minor cleanup (not part of this PR, but since you're modifying adjacent code): You can probably remove this `(void)clangDC;`.

Comment (zoecarver) [reply lib/AST/ASTMangler.cpp:1825]: Will do. I'll make it another PR (that way if for whatever reason we need to revert this one, the cleanup change will still exist).

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Review by varungandhi-apple — APPROVED** [review]

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please test Windows platform.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux
@swift-ci please smoke test OS X

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux

**Comment by zoecarver** [comment]
@swift-ci please smoke test OS X

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux
